File: Android_Components/Broadcast_Receiver/app/src/main/java/com/example/broadcast_receiver/MainActivity.java
package com.example.broadcast_receiver;

import androidx.appcompat.app.AppCompatActivity;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;

public class MainActivity extends AppCompatActivity implements View.OnClickListener{

    // Custom Broadcast Receiver parameters
    private static final String ACTION_CUSTOM_BROADCAST = BuildConfig.APPLICATION_ID + ".ACTION_CUSTOM_BROADCAST";
    private IntentFilter filter2 = new IntentFilter(ACTION_CUSTOM_BROADCAST);
    Button sendLocalBroadcast;

    // System Broadcast Receiver parameters
    private MyReceiver myReceiver = new MyReceiver();
    // Intent filters specify the types of intents a component can receive.
    private IntentFilter filter1 = new IntentFilter();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        sendLocalBroadcast = findViewById(R.id.sendBroadcast);
        sendLocalBroadcast.setOnClickListener(this);

        // When the system receives an Intent as a broadcast, it searches the broadcast receivers based on the action value specified in the IntentFilter object.
        filter1.addAction(Intent.ACTION_POWER_DISCONNECTED);
        filter1.addAction(Intent.ACTION_POWER_CONNECTED);
        filter1.addAction(Intent.ACTION_HEADSET_PLUG);
        filter1.addAction(Intent.ACTION_BATTERY_LOW);
        filter1.addAction(Intent.ACTION_AIRPLANE_MODE_CHANGED);

        // Registering receiver to receive System Broadcasts
        this.registerReceiver(myReceiver, filter1);

        // Registering receiver to receive Custom Broadcasts
        LocalBroadcastManager.getInstance(this).registerReceiver(myReceiver, filter2);
    }

    @Override
    protected void onDestroy() {
        // Unregistering from System Broadcast
        this.unregisterReceiver(myReceiver);
        // Unregistering from Custom Broadcast
        LocalBroadcastManager.getInstance(this).unregisterReceiver(myReceiver);

        super.onDestroy();
    }

    @Override
    public void onClick(View view) {
        // Make a custom intent object with String action as app's package name followed by custom action name
        Intent customBroadcastIntent = new Intent(ACTION_CUSTOM_BROADCAST);
        // Send the broadcast locally within the app
        LocalBroadcastManager.getInstance(this).sendBroadcast(customBroadcastIntent);
    }
}

File: Layouts/Affirmations_App/app/src/main/java/com/example/affirmations/MainActivity.kt
package com.example.affirmations

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.affirmations.adapter.AffirmationAdapter
import com.example.affirmations.data.DataSource
import com.example.affirmations.databinding.ActivityMainBinding
import com.example.affirmations.model.Affirmation

class MainActivity : AppCompatActivity() {

    private lateinit var _binding: ActivityMainBinding
    private val binding get() = _binding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(_binding.root)
//        setContentView(R.layout.activity_main)

        val myAffirmations: List<Affirmation> = DataSource().affirmations

        binding.recyclerView.adapter = AffirmationAdapter(this, myAffirmations)
        binding.recyclerView.layoutManager = LinearLayoutManager(this)
        binding.recyclerView.setHasFixedSize(true)

//        val recyclerView = findViewById<RecyclerView>(R.id.recycler_view)
//        recyclerView.adapter = AffirmationAdapter(this, myAffirmations)
//        recyclerView.layoutManager = GridLayoutManager(this,2)
//        recyclerView.setHasFixedSize(true)
    }
}